apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpNetworkTransport.kt

```diff
@@ -78,58 +80,92 @@ private constructor(
       // See https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md
-      val bodies: Flow<BufferedSource> = if (httpResponse.isMultipart) {
-        multipartBodyFlow(httpResponse)
+      if (httpResponse.isMultipart) {
+        emitAll(
+            multipleResponses(request.operation, customScalarAdapters, httpResponse)
+                .map { it.withHttpInfo(request.requestUuid, httpResponse, millisStart) }
+        )
       } else {
-        flowOf(httpResponse.body!!)
+        emit(
+            singleResponse(request.operation, customScalarAdapters, httpResponse)
+                .withHttpInfo(request.requestUuid, httpResponse, millisStart)
+        )
       }
+    }
+  }
 
-      emitAll(
-          bodies.map { body ->
-            // On native, body will be freezed in worker.doWork, but reading a BufferedSource is a mutating operation.
-            // So we read the bytes into a ByteString first, and create a new BufferedSource from it after freezing.
-            val bodyByteString = if (platform() == Platform.Native) body.readByteString() else null
-            // Do not capture request
-            val operation = request.operation
-            val response = try {
-              if (bodyByteString != null) {
-                // Native case, use bodyByteString
-                worker.doWork {
-                  operation.parseJsonResponse(
-                      jsonReader = Buffer().apply { write(bodyByteString) }.jsonReader(),
-                      customScalarAdapters = customScalarAdapters
-                  )
-                }
-              } else {
-                // Non-native case, use body
-                worker.doWork {
-                  operation.parseJsonResponse(
-                      jsonReader = body.jsonReader(),
-                      customScalarAdapters = customScalarAdapters
-                  )
-                }
-              }
-            } catch (e: Exception) {
-              throw wrapThrowableIfNeeded(e)
-            }
-
-            @Suppress("DEPRECATION")
-            response.newBuilder()
-                .requestUuid(request.requestUuid)
-                .addExecutionContext(
-                    HttpInfo(
-                        startMillis = millisStart,
-                        endMillis = currentTimeMillis(),
-                        statusCode = httpResponse.statusCode,
-                        headers = httpResponse.headers
-                    )
-                )
-                .build()
-          }
-      )
+  private suspend fun <D : Operation.Data> singleResponse(
+      operation: Operation<D>,
+      customScalarAdapters: CustomScalarAdapters,
+      httpResponse: HttpResponse,
+  ): ApolloResponse<D> {
+    val response = try {
+        worker.doWork {
+          operation.parseJsonResponse(
+              jsonReader = httpResponse.body!!.jsonReader(),
```

Comment:
How come we don't need the `val bodyByteString = if (platform() == Platform.Native) body.readByteString() else null` thing?

Reply:
This was needed when capturing/freezing `body`, but now we capture/freeze `httpResponse` so it can be simple again.

Reply:
But freezing `httpResponse` will deep freeze all its properties so it's the same thing no?

Reply:
Yes but `HttpResponse.body` is `get() = bodySource ?: bodyString?.let { Buffer().write(it) }` and in native we're in the `bodyString` case.

Reply:
👍  got it!